lib/inventory/controller.go

```diff
@@ -0,0 +1,302 @@
+// messages are processed by invoking the appropriate methods on the Auth interface.
+type Controller struct {
+	store           *Store
+	auth            Auth
+	serverKeepAlive time.Duration
+	serverTTL       time.Duration
+	testEvents      chan testEvent
+	closeContext    context.Context
+	cancel          context.CancelFunc
+}
+
+func NewController(auth Auth, opts ...controllerOption) *Controller {
+	var options controllerOpts
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	if options.serverKeepAlive == 0 {
+		baseKeepAlive := apidefaults.ServerKeepAliveTTL()
+		// use 1.5x standard server keep alive since we use a jitter that
+		// shortens the actual average interval.
+		options.serverKeepAlive = baseKeepAlive + (baseKeepAlive / 2)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	return &Controller{
+		store:           NewStore(),
+		serverKeepAlive: options.serverKeepAlive,
+		serverTTL:       apidefaults.ServerAnnounceTTL,
+		auth:            auth,
+		testEvents:      options.testEvents,
+		closeContext:    ctx,
+		cancel:          cancel,
+	}
+}
+
+// RegisterControlStream registers a new control stream with the controller.
+func (c *Controller) RegisterControlStream(stream client.UpstreamInventoryControlStream, hello proto.UpstreamInventoryHello) {
+	handle := newUpstreamHandle(stream, hello)
+	c.store.Insert(handle)
+	go c.handleControlStream(handle)
+}
+
+// GetControlStream gets a control stream for the given server ID if one exists (if multiple control streams
+// exist one is selected pseudorandomly).
+func (c *Controller) GetControlStream(serverID string) (handle UpstreamHandle, ok bool) {
+	handle, ok = c.store.Get(serverID)
+	return
+}
+
+// Iter iterates across all handles registered with this controller.
+// note: if multiple handles are registered for a given server, only
+// one handle is selected pseudorandomly to be observed.
+func (c *Controller) Iter(fn func(UpstreamHandle)) {
+	c.store.Iter(fn)
+}
+
+func (c *Controller) testEvent(event testEvent) {
+	if c.testEvents == nil {
+		return
+	}
+	c.testEvents <- event
+}
+
+// handleControlStream is the "main loop" of the upstream control stream. It handles incoming messages
+// and also manages keepalives for previously heartbeated state.
+func (c *Controller) handleControlStream(handle *upstreamHandle) {
+	c.testEvent(handlerStart)
+	defer func() {
+		c.store.Remove(handle)
+		handle.Close() // no effect if CloseWithError was called below
+		c.testEvent(handlerClose)
+	}()
+	keepAliveInterval := interval.New(interval.Config{
+		Duration:      c.serverKeepAlive,
+		FirstDuration: utils.HalfJitter(c.serverKeepAlive),
+		Jitter:        utils.NewSeventhJitter(),
+	})
+	defer keepAliveInterval.Stop()
+
+	for {
+		select {
+		case msg := <-handle.Recv():
+			switch m := msg.(type) {
+			case proto.UpstreamInventoryHello:
+				log.Warnf("Unexpected upstream hello on control stream of server %q.", handle.Hello().ServerID)
```

Comment:
I considered it, but didn't bother adding it since I couldn't think of any useful controller-level fields worth adding. Seemed to me that the ServerID associated with the specific stream was the "key" piece of information, and I didn't want to pre-allocate potentially many thousands of logger instances for that.  Decided it was simpler to just mention the ServerID of the associated stream in all the log lines.